torch/testing/_internal/common_fsdp.py

```diff
@@ -172,43 +263,97 @@ def get_loss(self, input, output):
     def run_backward(self, loss):
         loss.backward()
 
+    @staticmethod
+    def init(
+        group: dist.ProcessGroup,
+        fsdp_init_mode: FSDPInitMode,
+        cuda_init_mode: CUDAInitMode,
+        fsdp_kwargs: Optional[Dict[str, Any]] = None,
+        deterministic: bool = False,
+        add_bn: bool = True,
+    ) -> Union[nn.Module, FSDP]:
+        """
+        Initializes a :class:`TransformerWithSharedParams` instance.
+
+        Args:
+            fsdp_init_mode (FSDPInitMode): If ``NO_FSDP``, then does not wrap
+                any modules with FSDP. If ``RECURSIVE``, then wraps with
+                top-level FSDP. By default, the top-level FSDP uses the
+                ``transformer_auto_wrap_policy()`` for encoder and decoder
+                layers, but a different auto wrap policy may be specified via
+                ``fsdp_kwargs``.
+            cuda_init_mode (CUDAInitMode): Determines model movement to CUDA.
+            fsdp_kwargs (Optional[Dict[str, Any]]): Optional keyword arguments
+                forwarded to the FSDP constructor.
+            deterministic (bool): Whether to make the model deterministic
+                across constructions.
+            add_bn (bool): Whether to include batch norm in the model.
+        """
+        if fsdp_kwargs is None:
+            fsdp_kwargs = {}
+        if fsdp_init_mode == FSDPInitMode.NO_FSDP:
+            return TransformerWithSharedParams(group, cuda_init_mode, add_bn, deterministic)
+        elif fsdp_init_mode == FSDPInitMode.RECURSIVE:
+            if "cpu_offload" in fsdp_kwargs and fsdp_kwargs["cpu_offload"].offload_params:
+                assert cuda_init_mode == CUDAInitMode.CUDA_NEVER, f"{cuda_init_mode}"
+            # Default to the `transformer_auto_wrap_policy()`
```

Comment:
Now that we have the `transformer_auto_wrap_policy()`, it makes more sense to use this instead of always using a single top-level FSDP to wrap the `TransformerWithSharedParams`, which is not a meaningful wrapping.